### ISSUE TITLE
ssl-min-ver not respected in ssl-default-server-options or default-server

### ISSUE THREAD

**Issue by anitgandhi** [haproxy/haproxy#595]
Note that this was discussed on Slack/IRC and it was suggested to open a bug report here. 

## Output of `haproxy -vv` and `uname -a`

```
# haproxy -vv
HA-Proxy version 1.8.22-1~bpo9+1 2019/10/26
Copyright 2000-2019 Willy Tarreau <willy@haproxy.org>

Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-1.8.22=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-null-dereference -Wno-unused-label
  OPTIONS = USE_GETADDRINFO=1 USE_ZLIB=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_SYSTEMD=1 USE_PCRE2=1 USE_PCRE2_JIT=1 USE_NS=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with OpenSSL version : OpenSSL 1.1.0k  28 May 2019
Running on OpenSSL version : OpenSSL 1.1.0l  10 Sep 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2
Built with Lua version : Lua 5.3.3
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Encrypted password support via crypt(3): yes
Built with multi-threading support.
Built with PCRE2 version : 10.22 2016-07-29
PCRE2 library supports JIT : yes
Built with zlib version : 1.2.8
Running on zlib version : 1.2.8
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with network namespace support.

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace

# uname -a
Linux 2bff1152-3d67-47bb-b223-a9d970b75bc8 4.9.0-11-amd64 #1 SMP Debian 4.9.189-3+deb9u1 (2019-09-20) x86_64 GNU/Linux
```

## What's the configuration?

Config A (causes this issue):
```
global
  ...
  ssl-default-server-options ssl-min-ver TLSv1.1 no-tls-tickets
  ssl-server-verify none

defaults
  ...

frontend http_9000
  bind 0.0.0.0:9000
  default_backend http_9000_backend
  mode http
  ...

backend http_9000_backend
  mode http
  ...
  default-server ssl
  server node-a <IP-a>:443
  server node-b <IP-b>:443
```

Config B (causes this issue):
```
global
  ...
  ssl-default-server-options no-tls-tickets
  ssl-server-verify none

defaults
  ...
  default-server ssl-min-ver TLSv1.1

...

backend http_9000_backend
  ...
  default-server ssl
  server node-a <IP-a>:443
  server node-b <IP-b>:443
```

Config C (causes this issue):
```
global
  ...
  ssl-default-server-options no-tls-tickets
  ssl-server-verify none

defaults
  ...

...

backend http_9000_backend
  ...
  default-server ssl ssl-min-ver TLSv1.1
  server node-a <IP-a>:443 track 
  server node-b <IP-b>:443 track
```

Config D (workaround, does NOT cause this issue )
```
...

backend http_9000_backend
  ...
  default-server ssl
  server node-a <IP-a>:443 ssl-min-ver TLSv1.1
  server node-b <IP-b>:443 ssl-min-ver TLSv1.1
```

## Steps to reproduce the behavior

1. Create 2 backend servers (`node-a`, `node-b`) running a simple HTTPS web server, configured to only accept TLSv1.0 connections. In my testing, I used Nginx with its default server on 443.
2. Use the config above for HAProxy 1.8, 1.9, 2.0, or 2.1, replacing the IPs accordingly.
3. Run a `GET` request to `http://localhost:9000/`, and see a `200` response. 

## Actual behavior

I observed that a client `GET localhost:9000/` results in a `200` and I get the `Nginx` default `index.html` back, and that HAProxy negotiated a TLSv1.0 connection to the backend server (based on Nginx logs). 

## Expected behavior

I expected that the use of either `global.ssl-default-server-options ssl-min-ver TLSv1.1`, `defaults.default-server ssl-min-ver TLSv1.1`, or `backend.default-server ssl-min-ver TLSv1.1` would have resulted in HAProxy unable to create a connection to the backend, and the client would receive a 503. 

## Do you have any idea what may have caused this?

No, but if I were to guess, a missing link in the hierarchical config somewhere, such that the `server` line is not inheriting from `default-server` or `ssl-default-server-options`. 

## Do you have an idea how to solve the issue?

Currently the only workaround I can see is to use `ssl-min-ver TLSv1.1` on each `server` line in each `backend` section of my config. When I do so, I see the expected behavior described above.

**Comment by jmagnin** [comment]
it looks like the default values set with ssl-default-server-options for ssl-min-ver ssl-max-ver are only applied when check-ssl is set on the server line.

the following patch should fix the issue.

```
diff --git a/src/ssl_sock.c b/src/ssl_sock.c
index 376e62411..ddb495a33 100644
--- a/src/ssl_sock.c
+++ b/src/ssl_sock.c
@@ -10050,6 +10050,16 @@ static int srv_parse_ssl(char **args, int *cur_arg, struct proxy *px, struct ser
        if (global_ssl.connect_default_ciphersuites && !newsrv->ssl_ctx.ciphersuites)
                newsrv->ssl_ctx.ciphersuites = strdup(global_ssl.connect_default_ciphersuites);
 #endif
+       newsrv->ssl_ctx.options |= global_ssl.connect_default_ssloptions;
+       newsrv->ssl_ctx.methods.flags |= global_ssl.connect_default_sslmethods.flags;
+
+       if (!newsrv->ssl_ctx.methods.min)
+               newsrv->ssl_ctx.methods.min = global_ssl.connect_default_sslmethods.min;
+
+       if (!newsrv->ssl_ctx.methods.max)
+               newsrv->ssl_ctx.methods.max = global_ssl.connect_default_sslmethods.max;
+
+
        return 0;
 }
```

**Comment by anitgandhi** [comment]
I was facing the same issue when using `default-server ssl-min-ver TLSv1.1` via `defaults` and `backend` sections, so I think there would need to be cases for that as well.

**Comment by jmagnin** [comment]
I think my fix is incomplete. 

the following works:
```
global
  ssl-default-server-options ssl-max-ver TLSv1.1

listen in
  bind *:8080
  mode http
  server s1 some.ssl.site:443  ssl verify none
```
but 

```
global
  ssl-default-server-options ssl-max-ver TLSv1.1

listen in
  bind *:8080
  mode http
  default-server ssl verify none
  server s1 some.ssl.site:443  
```
does not apply ssl-max-ver to the server line.


edit: with this patch both configs behave as expected.

```
diff --git a/src/server.c b/src/server.c
index 4c745d655..f90cfff5a 100644
--- a/src/server.c
+++ b/src/server.c
@@ -1643,6 +1643,15 @@ static void srv_ssl_settings_cpy(struct server *srv, struct server *src)
                srv->ssl_ctx.verify_host = strdup(src->ssl_ctx.verify_host);
        if (src->ssl_ctx.ciphers != NULL)
                srv->ssl_ctx.ciphers = strdup(src->ssl_ctx.ciphers);
+       if (src->ssl_ctx.options)
+               srv->ssl_ctx.options = src->ssl_ctx.options;
+       if (src->ssl_ctx.methods.flags)
+               srv->ssl_ctx.methods.flags = src->ssl_ctx.methods.flags;
+       if (src->ssl_ctx.methods.min)
+               srv->ssl_ctx.methods.min = src->ssl_ctx.methods.min;
+       if (src->ssl_ctx.methods.max)
+               srv->ssl_ctx.methods.max = src->ssl_ctx.methods.max;
+
 #if (HA_OPENSSL_VERSION_NUMBER >= 0x10101000L && !defined OPENSSL_IS_BORINGSSL)
        if (src->ssl_ctx.ciphersuites != NULL)
                srv->ssl_ctx.ciphersuites = strdup(src->ssl_ctx.ciphersuites);
diff --git a/src/ssl_sock.c b/src/ssl_sock.c
index 376e62411..ddb495a33 100644
--- a/src/ssl_sock.c
+++ b/src/ssl_sock.c
@@ -10050,6 +10050,16 @@ static int srv_parse_ssl(char **args, int *cur_arg, struct proxy *px, struct ser
        if (global_ssl.connect_default_ciphersuites && !newsrv->ssl_ctx.ciphersuites)
                newsrv->ssl_ctx.ciphersuites = strdup(global_ssl.connect_default_ciphersuites);
 #endif
+       newsrv->ssl_ctx.options |= global_ssl.connect_default_ssloptions;
+       newsrv->ssl_ctx.methods.flags |= global_ssl.connect_default_sslmethods.flags;
+
+       if (!newsrv->ssl_ctx.methods.min)
+               newsrv->ssl_ctx.methods.min = global_ssl.connect_default_sslmethods.min;
+
+       if (!newsrv->ssl_ctx.methods.max)
+               newsrv->ssl_ctx.methods.max = global_ssl.connect_default_sslmethods.max;
+
+
        return 0;
 }
```

**Comment by anitgandhi** [comment]
I can confirm that this is fixed in 2.2-dev7. Using `default-server` in `defaults` or per `backend`, and the global `ssl-default-server-options` as well, all of them work with 2.2-dev7.
